[#3] feat: 회원가입 기능 구현

zeroooooowest opened the pull request:
데이터베이스는 로컬에서 docker 환경의 MySQL 5.7 버전을 사용했고 테스트시에는 H2 인메모리 DB 활용하였습니다.

user 스키마 sql 은 다음과 같습니다.(resources/schema.sql)
```sql
DROP TABLE IF EXISTS users;

CREATE TABLE users
(
    id       VARCHAR(255) PRIMARY KEY,
    email    VARCHAR(255) UNIQUE,
    password VARCHAR(255) NOT NULL,
    nickname VARCHAR(15) NOT NULL,
    tel      VARCHAR(11) NOT NULL,
    birthday DATE NOT NULL
);
```

테스트 코드에서 MockMvc를 통해 이슈(#3)에 있는 요구사항을 검증하였고 로컬 환경에서 postman 으로도 확인했습니다.

암호는 SHA-256 알고리즘을 통해 단방향 암호화를 하였습니다. (복호화 불가능)

========================================================
<br>

### 간단한 API 설명

#### 회원가입

<u>POST  /users</u> 요청 RequestBody 예시

```json
{
	"user":
    {
        "email": "test@test.com",
        "password": "test",
        "nickname": "test",
        "tel": "01011112222",
        "birthday": "2020-01-01"
    }
}
```

<u>JSON returned</u>

```json
{
    "user": {
        "id": "46b33aef-f60d-4bea-afb4-cf3c3c7a5125",
        "email": "test@test.com",
        "nickname": "test",
        "tel": "01011112222",
        "birthday": "2020-01-01"
    }
}
```



<u>Errors(중복) Returned</u>

```json
{
    "errors": {
        "email": [
            "duplicated email"
        ]
    }
}
```

<u>Errors(이메일 형식) Returned</u>

```json
{
    "errors": {
        "email": [
            "should be an email"
        ]
    }
}
```

<br>

### 간단한 코드 설명

#### 패키지 구성 이유

도메인에 관련된 코드가 도메인과 관련 없는 코드를 통해 확산되면 도메인에 관련된 코드를 확인하고 추적하기 힘들어집니다. 스터디에서 읽었던 DDD 책에서 설명하고 있는 LAYERED ARCHITECTURE 를 참고하여 패키지를 나누었습니다.

**![LAYERED_ARCHITECTURE](https://wikibook.co.kr/images/readit/20141002/table4-1.png)**

>**복잡한 프로그램을 여러 개의 계층으로 나눠라. 응집력 있고 오직 아래에 위치한 계층에만 의존하는 각 계층에서 설계를 발전시켜라. 표준 아키텍처 패턴에 따라 상위 계층과의 결합을 느슨하게 유지하라. 도메인 모델과 관련된 코드는 모두 한 계층에 모으고 사용자 인터페이스 코드나 애플리케이션 코드, 인프라스트럭처 코드와 격리하라. 도메인 객체(표현이나 저장, 애플리케이션 작업을 관리하는 등의 책임에서 자유로운)는 도메인 모델을 표현하는 것에만 집중할 수 있다. 이로써 모델은 진화를 거듭해 본질적인 업무 지식을 포착해서 해당 업무 지식이 효과를 발휘할 수 있을 만큼 풍부하고 명확해질 것이다.**

* 참고

  https://wikibook.co.kr/article/layered-architecture/

  https://dzone.com/articles/layered-architecture-is-good


<br>
<br>
또한, 의도치 않은 부수효과를 방지하기 위한 CQRS 패턴에서 차용하여 간단한 해결 방법을 택하여 읽기 모델과 쓰기 모델을 코드 차원에서 분리하였습니다.

<br>

![CQRS](https://docs.microsoft.com/ko-kr/azure/architecture/patterns/_images/command-and-query-responsibility-segregation-cqrs-basic.png)

* 참고

  https://docs.microsoft.com/ko-kr/azure/architecture/patterns/cqrs


<br>

f-lab-dev commented on src/main/java/dabang/star/cafe/api/UsersApi.java:
```diff
@@ -0,0 +1,88 @@
+@RestController
+@RequiredArgsConstructor
+public class UsersApi {
+    private final UserRepository userRepository;
```
이건 취향차이지만 클래스 선언 아래에 공백을 하나 넣으면 가독성을 향상시킬 수 있습니다~

zeroooooowest replied:
공백 줄을 넣을까 커밋하기전에 고민을 했는데 전체적으로 확인해서 클래스 가독성 수정하겠습니다.

추가한 의존성이 테스트 관련 의존성인데 재식님과 얘기를 통해 사용해보기로 하였습니다.

f-lab-dev commented on src/main/java/dabang/star/cafe/MyBatisConfig.java:
```diff
@@ -0,0 +1,44 @@
+import javax.sql.DataSource;
+
+@Configuration
+@EnableTransactionManagement
```
이 어노테이션을 달아주면 스프링 내부적으론 어떤 일이 발생하나요?

zeroooooowest replied:
EnableTransactionManagement 어노테이션을 달아주면 등록된 PlatformTransactionManager(PTM) 타입의 빈을 트랜잭션 매니저로 활용합니다. 즉, spring boot autoconfigure 를 통해 등록되는 DataSourceTransactionManager(PTM 의 구현)를 활용하게 됩니다. 트랜잭션 매니저는 AOP를 통해 자동으로 예외 발생시 롤백 처리를 해줍니다.

f-lab-dev commented on src/main/java/dabang/star/cafe/infrastructure/repository/MyBatisUserRepository.java:
```diff
@@ -0,0 +1,34 @@
+package dabang.star.cafe.infrastructure.repository;
+
+import dabang.star.cafe.domain.user.User;
```
`dabang.star.cafe` 패키지 명명 규칙은 어떻게 지으신건가요?

zeroooooowest replied:
일반적으로 회사의 경우 naver.com 이면 com.naver.~ 의 형태로 짓는 걸로 알고 있습니다.
일반적으로 dabang 이 통용되는 도메인이 아니기 때문에 고민했으나 실제 회사 운영 환경이 아니라 적당한 패키지명을 생각하지 못해 사용하였습니다.

f-lab-dev commented on src/main/java/dabang/star/cafe/infrastructure/service/SHA256EncryptService.java:
```diff
@@ -0,0 +1,31 @@
+            messageDigest.update(password.getBytes());
+            byte[] byteData = messageDigest.digest();
+
+            StringBuffer stringBuffer = new StringBuffer();
```
`StringBuilder`와 `StringBuffer`의 차이는 무엇일까요? 여기는 어떤 것을 사용하는게 좋을까요?

zeroooooowest replied:
String은 기본적으로 불변객체이기 때문에 더하기 연산을 하게 되면 반복적으로 임시 String 객체를 생성해내게 되고 GC의 대상이 되므로 성능적으로 불리합니다.

따라서 StringBuffer와 StringBuilder를 쓰게 되는데 차이는 스레드 세이프의 유무입니다.
기본적으로 스프링MVC는 멀티스레드 환경이기 때문에 StringBuffer 사용이 적당하다 생각했습니다.

f-lab-dev replied:
네 멀티스레드 환경인데 여기 보이는 `stringBuffer` 변수가 여러 스레드에서 공유될 가능성이 있을까요?

zeroooooowest replied:
다시 생각해보니 메서드 호출시 스택 프레임이 만들어지고 내부에 선언된 지역 변수이므로 공유되지 않을 거 같습니다.
StringBuilder 로 수정하겠습니다.

f-lab-dev commented on src/main/resources/application.yaml:
```diff
@@ -0,0 +1,14 @@
+    password: root
+  jackson:
+    deserialization:
+      UNWRAP_ROOT_VALUE: true
```
이런 옵션들을 왜 설정해주신건지 의도도 같이 주석에 넣어주시면 좋겠습니다~

zeroooooowest replied:
기본적으로 API 서버에 요청시 json root에 식별할 수 있는 정보를 넣는 걸로 가정하였기 때문에
root 는 역직렬화 시키지 않도록 하였습니다. 주석에 추가하겠습니다.

f-lab-dev commented on src/main/resources/application.yaml:
```diff
@@ -0,0 +1,14 @@
+spring:
+  datasource:
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    url: jdbc:mysql://172.17.0.2:3306/yslocaldb
```
`application.yaml`은 모든 환경에 적용되는 설정입니다~ `profile`을 나눠주는게 좋지 않을까요?

zeroooooowest replied:
인강에서 profile 관해 들을때 번거롭게 느껴져서 test/resources 하위에 application.yaml(더 높은 우선순위)을 재정의 하는 형태로 작성하였는데 한 번 알아보고 적용해보겠습니다.

f-lab-dev replied:
번거롭지만 해줘야하는 작업 중 하나입니다~ 알아봐주시면 좋겠네요

f-lab-dev commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -0,0 +1,47 @@
+    </insert>
+
+    <select id="findById" resultMap="user">
+        select id, email, password, nickname, tel, birthday
```
예약어 같은 경우는 대문자로 써주시면 가독성을 더 높힐 수 있습니다~

zeroooooowest replied:
넵 수정하겠습니다~

f-lab-dev commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,11 @@
+DROP TABLE IF EXISTS users;
+
+CREATE TABLE users
```
인덱스 설정은 추가로 걸게 없어서 따로 해주지 않으신걸까요??

zeroooooowest replied:
인덱스에 대해 깊게 공부해보지 않아서 공부해보겠습니다.

f-lab-dev commented on src/test/java/dabang/star/cafe/api/UsersApiTest.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Test
+    void 회원가입에_성공하는지_테스트() throws Exception {
```
이런 한글 설명은 `@DisplayName`을 사용해보시면 좋을 것 같네요~

zeroooooowest replied:
넵 수정하겠습니다~

f-lab-dev commented on src/test/java/dabang/star/cafe/api/UsersApiTest.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Test
+    void 중복된_이메일_가입시_에러를_보여주는지_테스트() throws Exception {
```
다른곳도 마찬가지지만 `중복된 이메일로 가입을 시도하면 에러가 발생한다` 식의 스펙문서를 정의하듯이 써주시면 좋을 것 같아요~

f-lab-dev reviewed (commented):
여기서 사용하시는 기술 스택은 재식님과 협의가 전부 된 점인지도 궁금하네요~

jaesika commented on src/main/java/dabang/star/cafe/MyBatisConfig.java:
```diff
@@ -0,0 +1,44 @@
+    public SqlSessionTemplate sqlSessionTemplate(SqlSessionFactory sqlSessionFactory) {
+        return new SqlSessionTemplate(sqlSessionFactory);
+    }
+}
```
저는 MyBatis-Spring-Boot-Starter가 자동설정으로 sqlSessionFactory나 sqlSessionTemplate을 자동으로 빈으로 등록해주는 것으로 알고있는데 사용자 정의 설정으로 달라지는점은 무엇인가요?

zeroooooowest replied:
초기 프로젝트 생성시 runtimeOnly로 h2 가 등록되어있어 자꾸 datasource 를 못찾아서 작성하였는데
testRuntimeOnly 로 수정하면서 이부분은 필요 없는 부분이 맞네요. 따로 특별히 커스텀하게 설정해준 부분도 없구요.

f-lab-dev commented on src/test/java/dabang/star/cafe/api/UsersApiTest.java:
```diff
@@ -76,7 +78,8 @@ public void setUp() throws Exception {
 
     @Test
-    void 중복된_이메일_가입시_에러를_보여주는지_테스트() throws Exception {
+    @DisplayName("중복된 이메일로 가입 요청시 에러를 보여줍니다.")
```
어떤 에러를 보여주는지 좀 더 스펙이 명확하면 좋겠습니다~

f-lab-dev commented on src/main/java/dabang/star/cafe/api/UsersApi.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private Map<String, Object> userResponse(UserData userData) {
+        return new HashMap<String, Object>() {{
```
1. 요청과 응답의 내용에 어떤 내용들이 들어가는지 명시적이지 않기 때문에 보통 `HashMap`을 요청과 응답의 형태로 권장하지는 않습니다~
2. 공통적인 응답의 형태를 만들고싶으신거라면 클래스로 빼는 등 공통적인 처리를 할 수 있도록 해주시면 좋을 것 같아요

zeroooooowest replied:
넵. 기존 에러 커스텀 직렬화 클래스(ErrorResourceSerializer.java)을 살짝 수정하고
application.yaml에 spring.serialization.WRAP_ROOT_VALUE: true 옵션을 주어서 직렬화시 공통적으로 루트 엘리먼트를 추가토록 했습니다.

f-lab-dev commented on src/main/java/dabang/star/cafe/api/UsersApi.java:
```diff
@@ -0,0 +1,89 @@
+    private final EncryptService encryptService;
+
+    @PostMapping("/users")
+    public ResponseEntity<?> createUser(@Valid @RequestBody RegisterParam registerParam,
```
제네릭을 넣어주신것은 좋으나 어떤 타입이 들어가는지 명확하면 좋겠습니다~

jaesika commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -0,0 +1,47 @@
+        SELECT id, email, password, nickname, tel, birthday
+        FROM users
+        WHERE email = #{email}
+    </select>
```
쿼리를 `UserMapper.xml`과 `UserReadService.xml`로 구분하셨는데 둘 다 `findById`와 `findByEmail`을 가지고 있습니다. 둘의 차이는 패스워드 유무에 따라 반환객체가 조금 달라보이는데 `UserMapper.xml`에서는 패스워드가 포함되어야 하는건가요? 
또한 유저에 대한 쿼리를 2개의 파일로 구분한 이유도 궁금합니다.

jaesika commented on src/main/java/dabang/star/cafe/api/UsersApi.java:
```diff
@@ -0,0 +1,81 @@
+                registerParam.getBirthday()
+        );
+        userRepository.save(user);
+        UserData userData = userQueryService.findById(user.getId()).get();
```
회원을 저장하고 다시 조회를 하는 것은 잘 저장되었다는 걸 확인하는 과정일까요?
`save`만으로도 저장이 성공적으로 되었을거라고 생각하는데 저장 후 반환타입을 UserData로 하는 것에 대해 어떻게 생각하시나요?

zeroooooowest replied:
위 질문과도 연결되는 부분이라 생각해서 한번에 쓰겠습니다.

post 요청을 통해 연속성을 갖는 엔터티(여기선 user)가 생성되기 때문에 엔터티의 식별성(유일성)을 가리키는 id 를 담아서 리턴해야 합니다. json 페이로드가 커서 문제가 될 수 있다면 id라도 반환을 하는데 이 경우 이상적인 practice 는 보통 생성된 리소스의 URI 값을 반환하는 거라고 하네요. 아마 이 부분은 로이필딩이 정의한 의미의 REST 를 만족하기 위해 HATEOAS 적인 측면에서 말하는 거 같습니다.

UserData를 반환하는 것 그 자체는 JPA 쓰실 때 DTO 를 반환하는 것과 유사하다고 생각됩니다. 
엔터티 자체는 상당히 많은 곳에서 사용될 것이므로 (지금 당장은 복잡하지 않지만) 엔터티 자체를 반환하게 되면 추후에 엔터티가 변하므로 API 스펙자체가 변경되는 문제가 있습니다.

모든 연산이 양분되는 것도 아니고 예외사항도 있지만 기본적으로 객체지향 프로그래밍에서 명령-쿼리 분리(CQS: Command Query Separation)는 SOLID 원칙에서 SRP와 OCP를 만족하기 위해서 원칙처럼 사용됩니다. Command는 데이터 변경을 가하는 행위고 Query는 데이터 조회라고 생각하시면 될 거 같습니다.

JPA에서는 트레이드오프를 고려하여 예외적으로 insert 문(명령)에 대해 id 를 리턴(쿼리)하고 MyBatis에서도 selectKey 옵션등으로 id를 받아올 수는 있습니다. 저는 읽기모델과 쓰기모델을 분리하는 CQRS를 적용하였기 때문에 따로 읽기모델에 대해 다시 조회를 했습니다. 

CQRS는 단지 연산 수준이었던 CQS를 개체나 시스템 수준에서 분리한 것일뿐 그 이상의 의미는 없고 이를 구현하는 방식은 정말 다양합니다. 아키텍처에 은탄환은 없으므로 비즈니스로직이 정말 간단한 서비스라면 읽기모델과 쓰기모델을 분리하는 것 자체가 오히려 코드를 구현하는데 있어서 오버헤드를 늘릴 수도 있습니다.

요즘 유행하는 MSA 환경에서는 명령 어플리케이션과 조회 어플리케이션을 따로 두기도 하지만 기본적으로 저희 프로젝트는 모놀리스입니다. 풀리퀘 본문에 있는 그림처럼 데이터 차원에서의 분리를 위해 테이블을 분리할 수도 있습니다만 필수는 아니라고 하여 같은 DB 안에서 명령과 쿼리를 같은 데이터베이스 내 같은 table을 사용했습니다. 이는 CQRS를 코드 레벨에서 적용하여 명령 쿼리 분리를 통해 객체지향 적인 코드를 작성하기 상대적으로 유리해집니다.

만약 명령 쿼리를 아예 다른 DB를 쓴다면 부하에 따라 확장성 있는 설계를 가져갈 수도 있고 쿼리 DB는 레디스 등을 통해 캐싱으로 성능을 향상시키는 방법도 있다고 합니다. 이 경우 데이터 정합성을 위해 명령 DB에서 쿼리 DB로 전파시켜야 하는데 카프카 등의 메시지 큐를 사용하기도 한다고 합니다. 이 부분은 최범균님 영상 링크 남깁니다.
> * https://www.youtube.com/watch?v=xf0kXMTFJm8
> * https://www.youtube.com/watch?v=H1IF3BUeFb8&feature=emb_title

f-lab-dev commented:
두 분의 사용하신 기술스택이 다른데 어떤 기술들을 사용할 것인가에 대한 협의도 진행해주시면 좋겠습니다~
아직 협의가 안되어있는 것으로 보여서요

zeroooooowest commented:
> 두 분의 사용하신 기술스택이 다른데 어떤 기술들을 사용할 것인가에 대한 협의도 진행해주시면 좋겠습니다~
> 아직 협의가 안되어있는 것으로 보여서요

어제 슬랙을 통해 간략히 얘기하고 오늘 오전10시부터 패키징 관련해서 지속적으로 구글밋으로 대화 나누었습니다.

패키징 관련해서 유쾌한 스프링방에서 검색 키워드를 추천받아서 'split by layer, split by feature' 로 검색한 결과
[packaging strategy](https://ssowonny.medium.com/package-by-feature-in-clean-architecture-projects-e14d25e3905e) 해당링크에 나오는 "Clean Architecture: Split by Layer, Package by Feature" 에 가깝게 본문의 기준대로 LAYER를 나누어 진행하기로 하였습니다. 재식님께서 패키징 리팩토링 작업을 하실 예정입니다. 

ORM의 도움 없이 DDD에서 말하는 모든 개념을 완전히 구현하기 현실적으로 어렵기 때문에 DDD의 개념을 완벽하게 프로젝트에 적용하는 것이 아니라 비즈니스와 인프라를 역할적인 측면에서 명확하게 계층적으로 나눈다는 점에서 차용하였습니다. 토비의 스프링 1권 p.815 쪽 그림과 유사한 계층형 구조라고 생각합니다.

f-lab-dev commented on src/main/java/dabang/star/cafe/api/request/RegisterParam.java:
```diff
@@ -0,0 +1,28 @@
+    @Email(message = "should be an email")
+    private String email;
+    @NotBlank(message = "can't be empty")
+    private String password;
```
이렇게 붙어있으면 가독성이 많이 떨어집니다~ 필드간에 개행을 넣어주시면 가독성을 향상시킬 수 있을 것 같네요~

f-lab-dev commented on src/main/java/dabang/star/cafe/domain/user/UserService.java:
```diff
@@ -0,0 +1,15 @@
+    private final UserRepository userRepository;
+
+    public boolean exists(User user) {
+        return userRepository.findByEmail(user.getEmail()).isPresent();
```
이거는 쿼리로 구현해도 될 것 같습니다~ 여기선 유저를 통째로 불러오는데 boolean값만 가져올 수 있게 하면 네트워크 대역폭을 조금이라도 절약할 수 있겠죠

f-lab-dev commented on src/test/java/dabang/star/cafe/infrastructure/repository/MyBatisUserRepositoryTest.java:
```diff
@@ -37,4 +37,11 @@ void saveUserSuccess() {
     }
+
+    @Test
+    @DisplayName("이메일로 저장여부 검색에 성공해야 합니다.")
```
존재하지 않는 케이스도 추가해주면 좋을 것 같네요~ 항상 실패하는 케이스를 염두에 두고 테스트를 작성해야합니다

f-lab-dev commented on src/test/java/dabang/star/cafe/infrastructure/repository/MyBatisUserRepositoryTest.java:
```diff
@@ -39,7 +39,7 @@ void saveUserSuccess() {
 
     @Test
-    @DisplayName("이메일로 저장여부 검색에 성공해야 합니다.")
+    @DisplayName("이메일로 존재 여부 검색에 성공해야 합니다.")
```
이것 또한 스펙에 맞게 `존재하는 이메일로 검색한다면 true가 리턴됩니다` 식으로 메소드의 스펙을 정의하는식으로 써주는게 좋습니다~
